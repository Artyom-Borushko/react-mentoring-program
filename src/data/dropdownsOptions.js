export const movieDropdownOptions = ['edit', 'delete'];

export const moviesSortDropdownOptions = ['release date', 'title'];
