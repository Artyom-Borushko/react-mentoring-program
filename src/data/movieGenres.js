
export const listOfGenres = ['all', 'documentary', 'comedy', 'horror', 'crime'];
