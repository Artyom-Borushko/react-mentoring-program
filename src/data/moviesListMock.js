export const listOfMoviesMock = [
    { src: '/images/movies-list/pulpFiction.png', title: "Pulp Fiction", releaseYear: "2004", relevantGenres: ['action', 'crime'], rating: '9.2', duration: '2h 22min', description: 'Lorem ipsum dolor sit amet. Eum autem quod aut dolorem voluptatum nam dolores sunt. Ea velit placeat ea corporis illo rem quod nulla. Et maxime placeat ut vero placeat et harum sequi. A Quis expedita id dolores sint id repellendus atque sit suscipit obcaecati vel quia galisum quo cumque internos!', id: "uniqueMovieId1" },
    { src: '/images/movies-list/bohemianRhapsody.png', title: "Bohemian Rhapsody", releaseYear: "2003", relevantGenres: ['documentary', 'drama'], rating: '8.4', duration: '1h 27min', description: 'Lorem ipsum dolor sit amet. Eum autem quod aut dolorem voluptatum nam dolores sunt. Ea velit placeat ea corporis illo rem quod nulla. Et maxime placeat ut vero placeat et harum sequi. A Quis expedita id dolores sint id repellendus atque sit suscipit obcaecati vel quia galisum quo cumque internos!', id: "uniqueMovieId2" },
    { src: '/images/movies-list/killBillVol2.png', title: "Kill Bill: Vol 2", releaseYear: "1994", relevantGenres: ['comedy', 'crime', 'oscar'], rating: '6.2', duration: '3h 12min', description: 'Lorem ipsum dolor sit amet. Eum autem quod aut dolorem voluptatum nam dolores sunt. Ea velit placeat ea corporis illo rem quod nulla. Et maxime placeat ut vero placeat et harum sequi. A Quis expedita id dolores sint id repellendus atque sit suscipit obcaecati vel quia galisum quo cumque internos!', id: "uniqueMovieId3" },
    { src: '/images/movies-list/avengersWarOfInfinity.png', title: "Avengers: War of Infinity", releaseYear: "2004", relevantGenres: ['action', 'adventure'], rating: '8.2', duration: '0h 56min', description: 'Lorem ipsum dolor sit amet. Eum autem quod aut dolorem voluptatum nam dolores sunt. Ea velit placeat ea corporis illo rem quod nulla. Et maxime placeat ut vero placeat et harum sequi. A Quis expedita id dolores sint id repellendus atque sit suscipit obcaecati vel quia galisum quo cumque internos!', id: "uniqueMovieId4" },
    { src: '/images/movies-list/inception.png', title: "Inception", releaseYear: "2003", relevantGenres: ['action', 'adventure', 'horror'], rating: '10.0', duration: '1h 39min', description: 'Lorem ipsum dolor sit amet. Eum autem quod aut dolorem voluptatum nam dolores sunt. Ea velit placeat ea corporis illo rem quod nulla. Et maxime placeat ut vero placeat et harum sequi. A Quis expedita id dolores sint id repellendus atque sit suscipit obcaecati vel quia galisum quo cumque internos!', id: "uniqueMovieId5" },
    { src: '/images/movies-list/reservoirDogs.png', title: "Reservoir dogs", releaseYear: "1994", relevantGenres: ['crime', 'oscar'], rating: '7.7', duration: '2h 00min', description: 'Lorem ipsum dolor sit amet. Eum autem quod aut dolorem voluptatum nam dolores sunt. Ea velit placeat ea corporis illo rem quod nulla. Et maxime placeat ut vero placeat et harum sequi. A Quis expedita id dolores sint id repellendus atque sit suscipit obcaecati vel quia galisum quo cumque internos!', id: "uniqueMovieId6" },
];
