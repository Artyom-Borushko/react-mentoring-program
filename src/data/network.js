export default 'http://localhost:4000/movies';
