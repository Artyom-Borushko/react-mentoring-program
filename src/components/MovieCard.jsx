import '../styles/movieCard.css';
import OptionsSelector from "./OptionsSelector";
import {movieDropdownOptions} from "../data/dropdownsOptions";

export default function MovieCard({title, releaseYear, relevantGenres, clickHandler, src}) {

    const handleDropdownSelection = (event) => {
        console.log('dropdown elements selected: ',
            event.target.innerText &&
            event.target.innerText.trim().toLowerCase()
        );
    };

    return (
        <div className={'movie-card'}>
            <OptionsSelector
                movieDropdownOptions={movieDropdownOptions}
                onDropdownSelection={handleDropdownSelection}
            />

            <img
                src={src}
                alt='movie card'
                aria-label={'movie card'}
                onClick={clickHandler}
            />
            <div className={'movie-name-year-wrapper'}>
                <p>{title}</p>
                <p className={'movie-release-year'}>{releaseYear}</p>
            </div>
            <p className={'movie-genres'}>{relevantGenres.join(', ')}</p>
        </div>
    );
}
