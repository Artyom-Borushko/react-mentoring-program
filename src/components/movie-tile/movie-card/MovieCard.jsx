import React, {memo} from "react";
import './movieCard.css';
import OptionsSelector from "../../shared/OptionsSelector";
import {movieDropdownOptions} from "../../../data/dropdownsOptions";
import {getMovieReleaseYear} from "../../../utilities/utilities";

export const MovieCard = memo(function MovieCard({movie, clickHandler, handleDropdownSelection}) {

    return (
        <div className={'movie-card'}>
            <OptionsSelector
                dropdownOptions={movieDropdownOptions}
                onDropdownSelection={(event) => handleDropdownSelection(event, movie)}
                controlImageSrc={'images/options-selector/threeDotsExpandOptionsIcon.png'}
                controlSource={'movie options'}
            />

            <img
                src={movie.src || '/images/movies-list/bohemianRhapsody.png'}
                alt='movie card'
                aria-label={'movie card'}
                onClick={() => clickHandler(movie)}
            />
            <div className={'movie-name-year-wrapper'}>
                <p>{movie.title}</p>
                <p className={'movie-release-year'}>{getMovieReleaseYear(movie.release_date)}</p>
            </div>
            <p className={'movie-genres'}>{movie.genres.join(', ')}</p>
        </div>
    );
})

export default MovieCard;
