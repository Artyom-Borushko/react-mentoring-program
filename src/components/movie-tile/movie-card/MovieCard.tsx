import React, { memo } from 'react';
import './movieCard.css';
import OptionsSelector from '../../shared/OptionsSelector';
import { movieDropdownOptions } from '../../../data/dropdownsOptions';
import { getMovieReleaseYear } from '../../../utilities/utilities';

interface Movie {
    src?: string;
    title: string;
    release_date: string;
    genres: string[];
}

interface ClickHandler {
    (movie: Movie): void;
}
interface HandleDropdownSelection {
    (event: React.SyntheticEvent, movie: Movie): void;
}

export const MovieCard = memo(({ movie, clickHandler, handleDropdownSelection }: {
    movie: Movie,
    clickHandler: ClickHandler,
    handleDropdownSelection: HandleDropdownSelection
}) => (
  <div className="movie-card">
    <OptionsSelector
      dropdownOptions={movieDropdownOptions}
      onDropdownSelection={(event) => handleDropdownSelection(event, movie)}
      controlImageSrc="images/options-selector/threeDotsExpandOptionsIcon.png"
      controlSource="movie options"
    />

    <img
      src={movie.src || '/images/movies-list/bohemianRhapsody.png'}
      alt="movie card"
      aria-label="movie card"
      onClick={() => clickHandler(movie)}
      role="presentation"
      data-testid="movie-card"
    />
    <div className="movie-name-year-wrapper">
      <p>{movie.title}</p>
      <p className="movie-release-year">{getMovieReleaseYear(movie.release_date)}</p>
    </div>
    <p className="movie-genres">{movie.genres.join(', ')}</p>
  </div>
));

export default MovieCard;
