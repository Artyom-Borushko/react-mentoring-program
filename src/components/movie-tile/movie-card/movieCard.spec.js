import {render, screen} from "@testing-library/react";
import React from "react";
import userEvent from "@testing-library/user-event";
import '@testing-library/jest-dom';
import MovieCard from "./MovieCard";
import {listOfMoviesMock} from "../../../data/moviesListMock";

describe("MovieCard component tests", () => {

    it("renders MovieCard component", () => {
        const {container} = render(<MovieCard
            movie={listOfMoviesMock[0]}
        />);

        expect(container).toMatchSnapshot();
    });

    it("event listener is working clicking on image", async () => {
        const mockClickHandler = jest.fn();
        render(<MovieCard
            movie={listOfMoviesMock[0]}
            clickHandler={mockClickHandler}
        />);

        await userEvent.click(screen.getByRole('img', {name: 'movie card'}));

        expect(mockClickHandler).toHaveBeenCalledTimes(1);
    });

    it("event listener is not working clicking on movie releaseYear", async () => {
        const mockClickHandler = jest.fn();
        render(<MovieCard
            movie={listOfMoviesMock[0]}
            clickHandler={mockClickHandler}
        />);

        await userEvent.click(screen.getByText('2018'));

        expect(mockClickHandler).toHaveBeenCalledTimes(0);
    });
});
