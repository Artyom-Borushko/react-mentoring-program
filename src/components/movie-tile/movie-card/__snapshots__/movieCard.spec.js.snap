// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MovieCard component tests renders MovieCard component 1`] = `
<div>
  <div
    class="movie-card"
  >
    <div>
      <img
        alt="expand options"
        aria-label="expand options"
        class="movie-expand-options-icon movie options"
        src="images/options-selector/threeDotsExpandOptionsIcon.png"
      />
    </div>
    <img
      alt="movie card"
      aria-label="movie card"
      src="/images/movies-list/pulpFiction.png"
    />
    <div
      class="movie-name-year-wrapper"
    >
      <p>
        Pulp Fiction
      </p>
      <p
        class="movie-release-year"
      >
        2004
      </p>
    </div>
    <p
      class="movie-genres"
    >
      action, crime
    </p>
  </div>
</div>
`;
